File: lib/timers.js
var cbs = {};
var ids = 0;

function createTimer(cb, timeout, c) {
  var id = ids++;
  cbs[id] = function () {
    cb();
    c--;
    if (c === 0) {
      delete cbs[id];
    } else {
      _native_set_timeout(id, timeout);
    }
  };

  _native_set_timeout(id, timeout);

  return id;
};

exports.setTimeout = function (cb, timeout) {
  return createTimer(cb, timeout, 1);
};

exports.setInterval = function (cb, interval) {
  return createTimer(cb, timeout, -1);
};

exports.clear = function (id) {
  delete cbs[id];
};

exports._native_callback = function (id) {
  var cb = cbs[id];
  cb && cb();
};

File: lib/timers.go
package lib

import (
	"fmt"
	"time"

	"github.com/dop251/goja"
)

func initTimers(vm *VM) {
	vm.Set("_native_set_timeout", func(call goja.FunctionCall) goja.Value {
		id := call.Argument(0).ToInteger()
		timeout := call.Argument(1).ToInteger()

		go func(id, timeout int64) {
			time.Sleep(time.Duration(timeout) * time.Millisecond)
			vm.RunString(fmt.Sprintf("require('timers')._native_callback(%d)", id))
		}(id, timeout)

		return vm.ToValue(nil)
	})

	vm.RunString(`
    var setTimeout = require('timers').setTimeout;
    var setInterval = require('timers').setInterval;
    var clearTimeout = require('timers').clear;
    var clearInterval = require('timers').clear;
  `)
}
